Collision detection maintenance

jngrad opened the pull request:
Description of changes:
- convert the Cython file `collision_detection.pyx` to a Python file
- reduce the complexity of the Python code by hiding implementation details in the C++ ScriptInterface class
- delegate the Python-to-C++ type conversion logic to the `AutoParameters` class instance
- bugfix: a runtime warning would print at random during collision due to undefined behavior
- fix broken error message: `glue_to_surface` and `bind_at_point_of_collision` require `VIRTUAL_SITES_RELATIVE`
- test all error messages
- calling `collision_detection.set_params()` with invalid arguments no longer leaves the system in an indeterminate state

jngrad committed:
core: Refactor collision detection exceptions

jngrad committed:
python: Rewrite collision detection ScriptInterface

Rely on the C++ AutoParameters class to remove all the type
conversion code from the Cython class. Convert the Cython file
to a Python file. Use the standard checkpointing mechanism.
Setting a protocol with invalid arguments no longer overwrites
the global state; the old protocol is automatically restored.

jngrad commented:
Note: this PR is only meant to make the code more maintainable for the 4.2.0 release. It also lays the groundwork for a possible overhaul in a future minor release, e.g. 4.3.0 (#4483).

jngrad commented:
ping

reinaual reviewed (dismissed):
I guess you are right with #4483...

reinaual commented on src/script_interface/collision_detection/CollisionDetection.hpp:
```diff
@@ -59,18 +97,97 @@ class CollisionDetection : public AutoParameters<CollisionDetection> {
+      auto const param_name = get_value<std::string>(kv.first);
+      if (expected_param_names.count(param_name) == 0) {
+        // throw if parameter is unknown
+        static_cast<void>(get_parameter(param_name));
```
this is probably violating the SRP...

kodiakhq[bot] committed:
Merge branch 'python' into coldet_refactor

jngrad committed:
core: Fix undefined behavior in collision detection

The uninitialized class member `bool exception_on_collision`
could take any value and would randomly trigger a runtime warning
or a fatal UBSAN report. The corresponding COLLISION_MODE_EXCEPTION
feature was removed in the Tcl version of ESPResSo in 2017.
